test race conditions

Saw some intermittent test failures trying to get https://github.com/git-lfs/git-lfs/pull/2353 out:

First, a simple go test failure. Kind of troubling that the logging outputs `second` twice:

```
--- FAIL: TestLoggerThrottlesLastWrite (0.02s)
	Error Trace:	log_test.go:149
	Error:		Not equal: "first\rsecond, done\n" (expected)
			        != "first\rsecond\rsecond, done\n" (actual)
```

Second has to do with the `migrate info (given unit)` integration test. I didn't capture the output unfortunately, but did trace it back to [this randomized repo name](https://github.com/git-lfs/git-lfs/blob/9f6ca0f22b431f020322c2bbbe4add2a12ecb2eb/test/test-migrate-fixtures.sh#L167). It returned a `/` at the front of the string, and something was trying to create that root directory.

Seeing similar issues trying to build with the `debian_7` docker img:

```
--- FAIL: TestLoggerThrottlesWrites (0.03s)
        Error Trace:    log_test.go:125
	Error:		Not equal: "first\rthird\rthird, done\n" (expected)
			        != "first\rsecond\rthird, done\n" (actual)

--- PASS: TestWaitingTaskIsThrottled (0.00s)
FAIL
FAIL	github.com/git-lfs/git-lfs/git/githistory/log	0.050s
```

@technoweenie thanks so much for opening this up. I have seen this a few times in CI, but never experienced it locally. I ran it through with `stress` and was able to observe the failure:

```
~/g/git-lfs (master) $ go test -c -race ./git/githistory/log -run TestLoggerThrottlesLastWrite

~/g/git-lfs (master) $ stress ./log.test -test.run=TestLoggerThrottlesLastWrite -test.cpu=10
16 runs so far, 0 failures
# ...
267 runs so far, 0 failures

/var/folders/23/1bpchtnj5y71fdyttct74cs00000gn/T/go-stress146920127
--- FAIL: TestLoggerThrottlesLastWrite (0.02s)
        Error Trace:    log_test.go:149
        Error:          Not equal: "first\rsecond, done\n" (expected)
                                != "first\rsecond\rsecond, done\n" (actual)

FAIL


ERROR: exit status 1


/var/folders/23/1bpchtnj5y71fdyttct74cs00000gn/T/go-stress843499794
--- FAIL: TestLoggerThrottlesLastWrite (0.02s)
        Error Trace:    log_test.go:149
        Error:          Not equal: "first\rsecond, done\n" (expected)
                                != "first\rsecond\rsecond, done\n" (actual)

FAIL


ERROR: exit status 1

289 runs so far, 2 failures
309 runs so far, 2 failures
^C
```

--- a/git/githistory/log/list_task.go
+++ b/git/githistory/log/list_task.go
@@ -1,6 +1,9 @@
 package log
 
-import "fmt"
+import (
+	"fmt"
+	"time"
+)
 
 // ListTask is a Task implementation that logs all updates in a list where each
 // entry is line-delimited.
@@ -11,24 +14,30 @@ import "fmt"
 //   msg: ..., done
 type ListTask struct {
 	msg string
-	ch  chan string
+	ch  chan *Update
 }
 
 // NewListTask instantiates a new *ListTask instance with the given message.
 func NewListTask(msg string) *ListTask {
 	return &ListTask{
 		msg: msg,
-		ch:  make(chan string, 1),
+		ch:  make(chan *Update, 1),
 	}
 }
 
 // Entry logs a line-delimited task entry.
 func (l *ListTask) Entry(update string) {
-	l.ch <- fmt.Sprintf("%s\n", update)
+	l.ch <- &Update{
+		S:  fmt.Sprintf("%s\n", update),
+		At: time.Now(),
+	}
 }
 
 func (l *ListTask) Complete() {
-	l.ch <- fmt.Sprintf("%s: ...", l.msg)
+	l.ch <- &Update{
+		S:  fmt.Sprintf("%s: ...", l.msg),
+		At: time.Now(),
+	}
 	close(l.ch)
 }
 
@@ -38,4 +47,4 @@ func (l *ListTask) Throttled() bool { return false }
 
 // Updates implements the Task.Updates function and returns a channel of updates
 // to log to the sink.
-func (l *ListTask) Updates() <-chan string { return l.ch }
+func (l *ListTask) Updates() <-chan *Update { return l.ch }

--- a/git/githistory/log/list_task_test.go
+++ b/git/githistory/log/list_task_test.go
@@ -12,7 +12,7 @@ func TestListTaskCallsDoneWhenComplete(t *testing.T) {
 
 	select {
 	case update, ok := <-task.Updates():
-		assert.Equal(t, "example: ...", update)
+		assert.Equal(t, "example: ...", update.S)
 		assert.True(t, ok,
 			"git/githistory/log: expected Updates() to remain open")
 	default:
@@ -36,7 +36,7 @@ func TestListTaskWritesEntries(t *testing.T) {
 	case update, ok := <-task.Updates():
 		assert.True(t, ok,
 			"git/githistory/log: expected ListTask.Updates() to remain open")
-		assert.Equal(t, "1\n", update)
+		assert.Equal(t, "1\n", update.S)
 	default:
 		t.Fatal("git/githistory/log: expected task.Updates() to have an update")
 	}

--- a/git/githistory/log/log.go
+++ b/git/githistory/log/log.go
@@ -107,7 +107,7 @@ func (l *Logger) enqueue(ts ...Task) {
 	if l == nil {
 		for _, t := range ts {
 			go func(t Task) {
-				for range <-t.Updates() {
+				for range t.Updates() {
 					// Discard all updates.
 				}
 			}(t)
@@ -196,17 +196,15 @@ func (l *Logger) logTask(task Task) {
 	logAll := !task.Throttled()
 	var last time.Time
 
-	var msg string
-	for msg = range task.Updates() {
-		now := time.Now()
-
-		if logAll || l.throttle == 0 || now.After(last.Add(l.throttle)) {
-			l.logLine(msg)
-			last = now
+	var update *Update
+	for update = range task.Updates() {
+		if logAll || l.throttle == 0 || update.At.After(last.Add(l.throttle)) {
+			l.logLine(update.S)
+			last = update.At
 		}
 	}
 
-	l.log(fmt.Sprintf("%s, done\n", msg))
+	l.log(fmt.Sprintf("%s, done\n", update.S))
 }
 
 // logLine writes a complete line and moves the cursor to the beginning of the

--- a/git/githistory/log/log_test.go
+++ b/git/githistory/log/log_test.go
@@ -9,25 +9,25 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-type ChanTask chan string
+type ChanTask chan *Update
 
-func (e ChanTask) Updates() <-chan string { return e }
+func (e ChanTask) Updates() <-chan *Update { return e }
 
 func (e ChanTask) Throttled() bool { return true }
 
-type UnthrottledChanTask chan string
+type UnthrottledChanTask chan *Update
 
-func (e UnthrottledChanTask) Updates() <-chan string { return e }
+func (e UnthrottledChanTask) Updates() <-chan *Update { return e }
 
 func (e UnthrottledChanTask) Throttled() bool { return false }
 
 func TestLoggerLogsTasks(t *testing.T) {
 	var buf bytes.Buffer
 
-	task := make(chan string)
+	task := make(chan *Update)
 	go func() {
-		task <- "first"
-		task <- "second"
+		task <- &Update{"first", time.Now()}
+		task <- &Update{"second", time.Now()}
 		close(task)
 	}()
 
@@ -43,16 +43,16 @@ func TestLoggerLogsTasks(t *testing.T) {
 func TestLoggerLogsMultipleTasksInOrder(t *testing.T) {
 	var buf bytes.Buffer
 
-	t1 := make(chan string)
+	t1 := make(chan *Update)
 	go func() {
-		t1 <- "first"
-		t1 <- "second"
+		t1 <- &Update{"first", time.Now()}
+		t1 <- &Update{"second", time.Now()}
 		close(t1)
 	}()
-	t2 := make(chan string)
+	t2 := make(chan *Update)
 	go func() {
-		t2 <- "third"
-		t2 <- "fourth"
+		t2 <- &Update{"third", time.Now()}
+		t2 <- &Update{"fourth", time.Now()}
 		close(t2)
 	}()
 
@@ -77,15 +77,15 @@ func TestLoggerLogsMultipleTasksWithoutBlocking(t *testing.T) {
 
 	l := NewLogger(&buf)
 	l.throttle = 0
-	t1, t2 := make(chan string), make(chan string)
+	t1, t2 := make(chan *Update), make(chan *Update)
 
 	l.widthFn = func() int { return 0 }
 	l.enqueue(ChanTask(t1))
 
-	t1 <- "first"
+	t1 <- &Update{"first", time.Now()}
 	l.enqueue(ChanTask(t2))
 	close(t1)
-	t2 <- "second"
+	t2 <- &Update{"second", time.Now()}
 	close(t2)
 
 	l.Close()
@@ -101,14 +101,14 @@ func TestLoggerLogsMultipleTasksWithoutBlocking(t *testing.T) {
 func TestLoggerThrottlesWrites(t *testing.T) {
 	var buf bytes.Buffer
 
-	t1 := make(chan string)
+	t1 := make(chan *Update)
 	go func() {
-		t1 <- "first"                     // t = 0     ms, throttle was open
-		time.Sleep(10 * time.Millisecond) // t = 10    ms, throttle is closed
-		t1 <- "second"                    // t = 10+ε  ms, throttle is closed
-		time.Sleep(10 * time.Millisecond) // t = 20    ms, throttle is open
-		t1 <- "third"                     // t = 20+ε  ms, throttle was open
-		close(t1)                         // t = 20+2ε ms, throttle is closed
+		start := time.Now()
+
+		t1 <- &Update{"first", start}                             // t = 0     ms, throttle was open
+		t1 <- &Update{"second", start.Add(10 * time.Millisecond)} // t = 10+ε  ms, throttle is closed
+		t1 <- &Update{"third", start.Add(20 * time.Millisecond)}  // t = 20+ε  ms, throttle was open
+		close(t1)                                                 // t = 20+2ε ms, throttle is closed
 	}()
 
 	l := NewLogger(&buf)
@@ -128,12 +128,13 @@ func TestLoggerThrottlesWrites(t *testing.T) {
 func TestLoggerThrottlesLastWrite(t *testing.T) {
 	var buf bytes.Buffer
 
-	t1 := make(chan string)
+	t1 := make(chan *Update)
 	go func() {
-		t1 <- "first"                     // t = 0     ms, throttle was open
-		time.Sleep(10 * time.Millisecond) // t = 10    ms, throttle is closed
-		t1 <- "second"                    // t = 10+ε  ms, throttle is closed
-		close(t1)                         // t = 10+2ε ms, throttle is closed
+		start := time.Now()
+
+		t1 <- &Update{"first", start}                             // t = 0     ms, throttle was open
+		t1 <- &Update{"second", start.Add(10 * time.Millisecond)} // t = 10+ε  ms, throttle is closed
+		close(t1)                                                 // t = 10+2ε ms, throttle is closed
 	}()
 
 	l := NewLogger(&buf)
@@ -156,11 +157,11 @@ func TestLoggerLogsAllDurableUpdates(t *testing.T) {
 	l.widthFn = func() int { return 0 }
 	l.throttle = 15 * time.Minute
 
-	t1 := make(chan string)
+	t1 := make(chan *Update)
 	go func() {
-		t1 <- "first"  // t = 0+ε  ms, throttle is open
-		t1 <- "second" // t = 0+2ε ms, throttle is closed
-		close(t1)      // t = 0+3ε ms, throttle is closed
+		t1 <- &Update{"first", time.Now()}  // t = 0+ε  ms, throttle is open
+		t1 <- &Update{"second", time.Now()} // t = 0+2ε ms, throttle is closed
+		close(t1)                           // t = 0+3ε ms, throttle is closed
 	}()
 
 	l.enqueue(UnthrottledChanTask(t1))

--- a/git/githistory/log/percentage_task.go
+++ b/git/githistory/log/percentage_task.go
@@ -3,6 +3,7 @@ package log
 import (
 	"fmt"
 	"sync/atomic"
+	"time"
 )
 
 // PercentageTask is a task that is performed against a known number of
@@ -17,14 +18,14 @@ type PercentageTask struct {
 	total uint64
 	// ch is a channel which is written to when the task state changes and
 	// is closed when the task is completed.
-	ch chan string
+	ch chan *Update
 }
 
 func NewPercentageTask(msg string, total uint64) *PercentageTask {
 	p := &PercentageTask{
 		msg:   msg,
 		total: total,
-		ch:    make(chan string, 1),
+		ch:    make(chan *Update, 1),
 	}
 	p.Count(0)
 
@@ -49,11 +50,14 @@ func (c *PercentageTask) Count(n uint64) (new uint64) {
 		percentage = 100 * float64(new) / float64(c.total)
 	}
 
-	msg := fmt.Sprintf("%s: %3.f%% (%d/%d)",
-		c.msg, percentage, new, c.total)
+	u := &Update{
+		S: fmt.Sprintf("%s: %3.f%% (%d/%d)",
+			c.msg, percentage, new, c.total),
+		At: time.Now(),
+	}
 
 	select {
-	case c.ch <- msg:
+	case c.ch <- u:
 	default:
 		// Use a non-blocking write, since it's unimportant that callers
 		// receive all updates.
@@ -69,7 +73,7 @@ func (c *PercentageTask) Count(n uint64) (new uint64) {
 // Updates implements Task.Updates and returns a channel which is written to
 // when the state of this task changes, and closed when the task is completed.
 // has been completed.
-func (c *PercentageTask) Updates() <-chan string {
+func (c *PercentageTask) Updates() <-chan *Update {
 	return c.ch
 }
 

--- a/git/githistory/log/percentage_task_test.go
+++ b/git/githistory/log/percentage_task_test.go
@@ -9,12 +9,12 @@ import (
 func TestPercentageTaskCalculuatesPercentages(t *testing.T) {
 	task := NewPercentageTask("example", 10)
 
-	assert.Equal(t, "example:   0% (0/10)", <-task.Updates())
+	assert.Equal(t, "example:   0% (0/10)", (<-task.Updates()).S)
 
 	n := task.Count(3)
 	assert.EqualValues(t, 3, n)
 
-	assert.Equal(t, "example:  30% (3/10)", <-task.Updates())
+	assert.Equal(t, "example:  30% (3/10)", (<-task.Updates()).S)
 }
 
 func TestPercentageTaskCalculatesPercentWithoutTotal(t *testing.T) {
@@ -23,7 +23,7 @@ func TestPercentageTaskCalculatesPercentWithoutTotal(t *testing.T) {
 	select {
 	case v, ok := <-task.Updates():
 		if ok {
-			assert.Equal(t, "example: 100% (0/0)", v)
+			assert.Equal(t, "example: 100% (0/0)", v.S)
 		} else {
 			t.Fatal("expected channel to be open")
 		}
@@ -37,7 +37,7 @@ func TestPercentageTaskCallsDoneWhenComplete(t *testing.T) {
 	select {
 	case v, ok := <-task.Updates():
 		if ok {
-			assert.Equal(t, "example:   0% (0/10)", v)
+			assert.Equal(t, "example:   0% (0/10)", v.S)
 		} else {
 			t.Fatal("expected channel to be open")
 		}
@@ -45,7 +45,7 @@ func TestPercentageTaskCallsDoneWhenComplete(t *testing.T) {
 	}
 
 	assert.EqualValues(t, 10, task.Count(10))
-	assert.Equal(t, "example: 100% (10/10)", <-task.Updates())
+	assert.Equal(t, "example: 100% (10/10)", (<-task.Updates()).S)
 
 	if _, ok := <-task.Updates(); ok {
 		t.Fatalf("expected channel to be closed")

--- a/git/githistory/log/task.go
+++ b/git/githistory/log/task.go
@@ -1,15 +1,25 @@
 package log
 
+import "time"
+
 // Task is an interface which encapsulates an activity which can be logged.
 type Task interface {
 	// Updates returns a channel which is written to with the current state
 	// of the Task when an update is present. It is closed when the task is
 	// complete.
-	Updates() <-chan string
+	Updates() <-chan *Update
 
 	// Throttled returns whether or not updates from this task should be
 	// limited when being printed to a sink via *log.Logger.
 	//
 	// It is expected to return the same value for a given Task instance.
 	Throttled() bool
+}
+
+// Update is a single message sent (S) from a Task at a given time (At).
+type Update struct {
+	// S is the message sent in this update.
+	S string
+	// At is the time that this update was sent.
+	At time.Time
 }

--- a/git/githistory/log/waiting_task.go
+++ b/git/githistory/log/waiting_task.go
@@ -1,18 +1,24 @@
 package log
 
-import "fmt"
+import (
+	"fmt"
+	"time"
+)
 
 // WaitingTask represents a task for which the total number of items to do work
 // is on is unknown.
 type WaitingTask struct {
 	// ch is used to transmit task updates.
-	ch chan string
+	ch chan *Update
 }
 
 // NewWaitingTask returns a new *WaitingTask.
 func NewWaitingTask(msg string) *WaitingTask {
-	ch := make(chan string, 1)
-	ch <- fmt.Sprintf("%s: ...", msg)
+	ch := make(chan *Update, 1)
+	ch <- &Update{
+		S:  fmt.Sprintf("%s: ...", msg),
+		At: time.Now(),
+	}
 
 	return &WaitingTask{ch: ch}
 }
@@ -24,7 +30,7 @@ func (w *WaitingTask) Complete() {
 
 // Done implements Task.Done and returns a channel which is closed when
 // Complete() is called.
-func (w *WaitingTask) Updates() <-chan string {
+func (w *WaitingTask) Updates() <-chan *Update {
 	return w.ch
 }
 

--- a/git/githistory/log/waiting_task_test.go
+++ b/git/githistory/log/waiting_task_test.go
@@ -9,7 +9,7 @@ import (
 func TestWaitingTaskDisplaysWaitingStatus(t *testing.T) {
 	task := NewWaitingTask("example")
 
-	assert.Equal(t, "example: ...", <-task.Updates())
+	assert.Equal(t, "example: ...", (<-task.Updates()).S)
 }
 
 func TestWaitingTaskCallsDoneWhenComplete(t *testing.T) {
@@ -18,7 +18,7 @@ func TestWaitingTaskCallsDoneWhenComplete(t *testing.T) {
 	select {
 	case v, ok := <-task.Updates():
 		if ok {
-			assert.Equal(t, "example: ...", v)
+			assert.Equal(t, "example: ...", v.S)
 		} else {
 			t.Fatal("expected channel to be open")
 		}

